### ISSUE TITLE
JetEngine: Ability to store Cache API results locally

### ISSUE THREAD

**Issue by jamsheehan** [CrocoBlock/suggestions#3482]
The REST API Endpoints functionality is a powerful piece of kit, however if you want to query the results to combine them with your local results you are quite limited. 

Let's say we have a client who has an API with their users details who were exported from their WP site, and managed some information externally to WordPress. So we can pull the data in, but we can't put it anywhere (using just JetEngine and WP). It would be great if we created a table based on the cached results in the local database. 

### Requirements
Add a toggle that would allow the user to store the retrieved data in a new table for the cache duration specified. This should come with a warning that the data can fill up the users database and will slow down the site when the update is running in addition to data may not be available during the update. So we should also define a limit of the amount of records stored, e.g. 5000, and a time to run the cached update. 

When you create the new save, the data will be stored in your DB in the following format: 

tablename_jet_rest_cache_<queryname>

The table columns generated would be based on the results returned based on the item path so if there's a returned value e.g.: 

```
 "avatar_urls": {
            "24": "https://secure.gravatar.com/avatar/xxx?s=24&d=retro&r=g",
            "48": "https://secure.gravatar.com/avatar/xxx?s=48&d=retro&r=g",
            "96": "https://secure.gravatar.com/avatar/xxx?s=96&d=retro&r=g"
        },
```
The results 24,48,96 should be saved as an array under avatar_urls. 

I think the site admin should be able to select the data types for the storage (so a key with the data type) and which ones to store, so when you click on a preview button it would display something like the following table will be created: 

| Store | Name | Data Type
| ------------- | ------------- | ------------- |
|  [x ]  | jet_cache_id |  bigint(20) unsigned
|  [ ]  | post_id  |  bigint(20) unsigned
|  [ ]  | meta_key |  varchar(255) unsigned
|  [ ]  | meta_value |  longtext

There should also be a summary of how many rows will be created e.g. 

This action will create 3,812 rows. Click here to to confirm: [ ]

The user should then have to assign the SQL types, we could add validation around specific phrases like post_id and make educated guesses what the type should be but allow the user to override. 

Once the endpoint is deleted the table is also deleted. 

As with anything we should also sanitise any data returned and stored.. 

The data will then be able to be queried by the Query Builder. This would allow the site admin to use the Query Builder to combine the stored data with local data using the SQL Query or just as a post type using the posts query if the cache is created as a post type. 

It could also be possible in a future update to specify timeouts for storing the data, e.g. 60 seconds, 180s etc (depending on the max timeout of the Wordpress installation). 

The ability to store the data would take JE Rest API's to the next level, and we could hook up the Smart Filters into the newly created table to enhance it further. 

Please feel free to reach out if you have any questions.

**Comment by tylertyza** [comment]
This is an amazing idea!

If the data was stored this way it would open up more opportunities... for example if you wanted to pull data from an API into a form field you could, meaning you could possibly prefill all the data from another website then push that data as an update to that website, letting you edit/view external data from one site. :D

**Comment by jamsheehan** [comment]
> This is an amazing idea!
> 
> If the data was stored this way it would open up more opportunities... for example if you wanted to pull data from an API into a form field you could, meaning you could possibly prefill all the data from another website then push that data as an update to that website, letting you edit/view external data from one site. :D

Yeah for my particular use case I've been using [Postman ](https://www.postman.com/) to achieve this, so what it does is grab data from Site A, do what I need to do by manipulating the data to match the Custom Content Type format, and then submit to Site B using the Crocoblock API into a Custom Content Type. So at 2am for the client the site overwrites the data on Site B and reconstructs it with the latest info from postman. It's not by far a perfect solution, but it works (_generally_) and keeps the client on budget.

**Comment by tylertyza** [comment]
> > This is an amazing idea!
> > If the data was stored this way it would open up more opportunities... for example if you wanted to pull data from an API into a form field you could, meaning you could possibly prefill all the data from another website then push that data as an update to that website, letting you edit/view external data from one site. :D
> 
> Yeah for my particular use case I've been using [Postman ](https://www.postman.com/) to achieve this, so what it does is grab data from Site A, do what I need to do by manipulating the data to match the Custom Content Type format, and then submit to Site B using the Crocoblock API into a Custom Content Type. So at 2am for the client the site overwrites the data on Site B and reconstructs it with the latest info from postman. It's not by far a perfect solution, but it works (_generally_) and keeps the client on budget.

Ahh okay, sounds like that would work. So you would be able to pull the data from the CCT into a form then edit it then push it back to your external source, but the changes would reflect the next day.

**Comment by jamsheehan** [comment]
Yep, the only reason I'm using postman for my particular case is because CB don't support the functionality (yet) and they have generous free tiers for API calls.. I could technically do more frequent updates but the free tier to postman is 1000 calls per month, so want to limit what it can do in case that drops to say 100 a month for the sake of my client. 

Ideally I'd do it all through a CB solution to reduce points of failure risk.
